extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -190,6 +201,60 @@ export class AmpAdExit extends AMP.BaseElement {
       this.user().error(TAG, 'Invalid JSON config', e);
       throw e;
     }
+
+    const ampAdResourceId = getAmpAdResourceId(this.element, this.win.top);
+    if (!ampAdResourceId) {
+      this.user().error(TAG, 'No friendly parent amp-ad element was found' +
+          ' for amp-ad-exit tag.');
+    }
+
+    this.unlisten_ = listen(this.getAmpDoc().win, 'message', event => {
+      const responseMessage = deserializeMessage(getData(event));
+
+      this.assertValidResponseMessage(responseMessage,
+          /** @type {string} */ (ampAdResourceId), event.origin);
+
+      this.vendorResponses_[responseMessage['vendor']] =
+          responseMessage['message'];
+    });
+  }
+
+  /** @override */
+  detachedCallback() {
+    if (this.unlisten_) {
+      this.unlisten_();
+    }
+  }
+
+  /**
+   *
+   * @param {?JsonObject|undefined} responseMessage The response object to validate.
+   * @param {string} expectedCreativeId The resource ID of the enclosing AMP ad (which
+   *     responseMessage['creativeId'] should match.
+   * @param {string} expectedVendor The 3p analytics vendor, which
+   *     responseMesssage['vendor'] should match.
+   */
+  assertValidResponseMessage(responseMessage, expectedCreativeId,
+                             expectedVendor) {
+    if (!responseMessage || !responseMessage['type'] ||
+      responseMessage['type'] != MessageType.IFRAME_TRANSPORT_RESPONSE ||
+      !responseMessage['creativeId'] ||
+      responseMessage['creativeId'] != expectedCreativeId) {
+      return;
+    }
+    dev().assert(responseMessage && responseMessage['message'],
+        'Received empty response from 3p analytics frame');
+    dev().assert(responseMessage['type'] &&
+      responseMessage['type'] == MessageType.IFRAME_TRANSPORT_RESPONSE,
+        'Received response message of invalid type from 3p analytics frame');
+    dev().assert(responseMessage['creativeId'] &&
+      responseMessage['creativeId'] == expectedCreativeId,
+        'Received malformed message from 3p analytics frame: ' +
+        'creativeId missing');
+    dev().assert(responseMessage['vendor'],
+        'Received malformed message from 3p analytics frame: ' +
+      'vendor missing');
```

Comment:
This doesn't look indented right.

Reply:
I agree. I think "gulp lint" may have an error with checks for functions that spill onto a 3rd line (or beyond), because I ran "gulp lint" and "gulp lint --fix" several times during the course of my own self-review...
Anyway, thanks, will fix.

Reply:
Fixed.